src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -48,8 +48,9 @@ internal class CompletionSource : ForegroundThreadAffinitizedObject, IAsyncCompl
 
         private static readonly EditorOptionKey<bool> NonBlockingCompletionEditorOption = new EditorOptionKey<bool>(NonBlockingCompletion);
 
-        private static readonly ConditionalWeakTable<RoslynCompletionItem, VSCompletionItem> s_roslynItemToVsItem =
-            new ConditionalWeakTable<RoslynCompletionItem, VSCompletionItem>();
+        // Use CWT to cache data needed to create VSCompletionItem, so the table would be cleared when Roslyn completion item cache is cleared.
+        private static readonly ConditionalWeakTable<RoslynCompletionItem, VSCompletionItemData> s_roslynItemToVsItemData =
+            new ConditionalWeakTable<RoslynCompletionItem, VSCompletionItemData>();
```

Comment:
it's a real pity this is such a fine-grained collection (CWTs are not cheap).  that said, this is likely a more than good enough solutoin for 16.1.

Reply:
An alternative would be adding an internal property bag (`ImmutableDictionary<string, object>`) for Roslyn completion item.

Reply:
that woudl also box on the object side.

my preference really is a `dictionary<item,itemdata>` where `itemdata` is  struct.  But then, the CWT needs to be keyed off of something that holds onto the entire dictionary.  it's doable, but requires more changes.  so i would hold off on it.

Reply:
I'd argue that box (or creating the VSCompletionItemData object if we just keep them as class) wouldn't be a problem since it's only done once for each cache item, and this would let us avoid using CWT